docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+   stack on them.
+2. For any shoot, enables the user to select an alternative ShootKapi scaling mode which maximises reliability and is
+   known to work well even under extreme API request profiles.
+
+The following approach is envisioned (discussion is limited to scaling ShootKapi):
+- The default scaling mode for `testing` shoots consists of a fixed number of 2 replicas, scaled by VPA. 
+- The default scaling mode for all other shoots, is the one proposed by this GEP.
+- For all shoots, the user is enabled to select an alternative scaling mode. In this mode, the user specifies a fixed
+  vertical scale, and a horizontal scale in the form of (MinimumReplicaCount, MaximumReplicaCount). HPA is used
+  to apply the horizontal scale.
+
+## Discussion and Limitations
+Overall, multidimensional autoscaling is a job which cannot be accomplished efficiently with the one-dimensional
+autoscalers currently available in the K8s ecosystem. The proposed approach is a compromise which relies on estimating
+actual compute resource demand based on an application metric (API request rate). Such estimate is inherently inaccurate,
+creating a need for a safety margin (excess replicas), and a risk of resource exhaustion under highly unusual application
+loads.
+
+#### Prometheus-metrics-adapter precludes the use of other sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot control-plane prometheus and aggregate prometheus) into
+the shoot cluster control loop. This has major implications on reliability. The alternative is a small, custom-built
+component registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and
+outputs request rate.
+
+The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
+custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  
+
+The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
+latency is deemed satisfactory.
+
+While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
+substitution with a signal path based on the aforementioned simple custom component. Such custom-built component,
+also has the advantage that it could easily coexist with other metrics providers by selectively forwarding queries to
+them. This supports future changes based on accumulated field experience, or evolving solution needs. 
+
+#### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads
```

Comment:
This section seems to be too detailed/confusing for an implementation proposal.

Reply:
Another reviewer requested that it be added. I don't have a strong opinion either way, but I guess it would be easier to converge toward consensus on the grounds "leave it in, those who don't need it can ignore it".